src/backend/partaj/core/api/referral.py

```diff
@@ -96,41 +98,96 @@ def get_permissions(self):
+        """
+        Update an existing referral.
+        """
+        referral = self.get_object()
+
+        referral.context = request.data.get("context")
+        referral.context = request.data.get("question")
+        referral.object = request.data.get("object")
+        referral.prior_work = request.data.get("prior_work")
+        referral.urgency_explanation = request.data.get("urgency_explanation")
+
+        if request.data.get("topic"):
+            try:
+                topic = models.Topic.objects.get(id=request.POST["topic"])
+                referral.topic = topic
+            except models.Topic.DoesNotExist:
+                return Response(
+                    status=400,
+                    data={
+                        "urgency_level": [
+                            f"{request.data.get('Topic')} is not a valid topic."
+                        ]
+                    },
+                )
+
+        referral.users.clear()
+        referral.users.set(request.data.getlist("users"))
+
+        if request.data.get("urgency_level"):
             # Do not create the referral until we can completely validate it: we need to first
             # make sure the urgency ID we received matches an existing urgency level.
             try:
                 referral_urgency = models.ReferralUrgency.objects.get(
-                    id=request.POST["urgency_level"]
+                    id=request.data.get("urgency_level")
                 )
             except models.ReferralUrgency.DoesNotExist:
                 return Response(
                     status=400,
                     data={
                         "urgency_level": [
-                            f"{request.POST['urgency_level']} is not a valid referral urgency id."
+                            f"{request.data.get('urgency_level')} is not a valid referral urgency id."
                         ]
                     },
                 )
-
-            # Create the referral from existing data
-            referral = form.save()
-
             # Add in the urgency level we found and the relevant unit
             referral.urgency_level = referral_urgency
+
+        referral.save()
+
+        # Create Attachment instances for the related files
+        files = request.FILES.getlist("files")
+        for file in files:
+            referral_attachment = models.ReferralAttachment(
+                file=file, referral=referral
+            )
+            referral_attachment.save()
+
+        return Response(data=ReferralSerializer(referral).data)
+
+    @action(
+        detail=True,
+        methods=["post"],
+        permission_classes=[UserIsReferralRequester],
+    )
+    # pylint: disable=invalid-name
+    def send(self, request, pk):
+        """
+        Update and Send an existing referral.
+        """
+        instance = self.get_object()
+
+        form = ReferralForm(
+            {
+                **{key: value for key, value in request.data.items()},
```

Comment:
This is an action path, not a POST. IMO we should fill the `ReferralForm` from data already on the Referral to just check it conforms to our standards. Then just call a transition. We don't want to pass the whole referral in the body of an action request.